public/app/features/explore/ContentOutline/ContentOutlineContext.tsx

```diff
@@ -21,30 +22,55 @@ const ContentOutlineContext = createContext<ContentOutlineContextProps | undefin
 export const ContentOutlineContextProvider = ({ children }: { children: ReactNode }) => {
   const [outlineItems, setOutlineItems] = useState<ContentOutlineItemContextProps[]>([]);
 
-  const register: RegisterFunction = useCallback(({ title, icon, ref }) => {
-    const id = uniqueId(`${title}-${icon}_`);
+  const register: RegisterFunction = useCallback(({ panelId, title, icon, ref }) => {
+    const id = uniqueId(`${panelId}-${title}-${icon}_`);
 
     setOutlineItems((prevItems) => {
-      const updatedItems = [...prevItems, { id, title, icon, ref }];
-
-      return updatedItems.sort((a, b) => {
-        if (a.ref && b.ref) {
-          const diff = a.ref.compareDocumentPosition(b.ref);
-          if (diff === Node.DOCUMENT_POSITION_PRECEDING) {
-            return 1;
-          } else if (diff === Node.DOCUMENT_POSITION_FOLLOWING) {
-            return -1;
-          }
-        }
-        return 0;
-      });
+      const matchIndex = prevItems.findIndex((item) => item.panelId === panelId);
+      const match = prevItems[matchIndex];
+
+      const matchIsParent = match && match.children;
+
+      let updatedItems = [...prevItems];
+
+      if (matchIsParent) {
+        match.children?.push({ id, panelId, title, icon, ref });
+        match.children?.sort(sortElementsByDocumentPosition);
+      } else if (match) {
+        const parent = {
+          ...match,
+          id: `section-${id}_`,
+          title: panelId,
+          children: [match, { id, panelId, title, icon, ref }],
+        };
+        parent.children.sort(sortElementsByDocumentPosition);
+        updatedItems.splice(matchIndex, 1, parent);
+      } else {
+        updatedItems = [...prevItems, { id, panelId, title, icon, ref }];
+        updatedItems.sort(sortElementsByDocumentPosition);
+      }
+
+      return updatedItems;
     });
 
     return id;
   }, []);
 
   const unregister = useCallback((id: string) => {
-    setOutlineItems((prevItems) => prevItems.filter((item) => item.id !== id));
+    setOutlineItems(
+      (prevItems) =>
+        prevItems
+          .filter((item) => item.id !== id)
+          .map((item) => {
+            if (item.children) {
+              item.children = item.children.filter((child) => child.id !== id);
+            }
+            return item;
+          })
+
+      //TODO: need logic here to remove section if it only has one child
```

Comment:
Would it be simpler if we always created the parent and just decided in the UI how to render it? (i.e. no nesting with single child item). 🤔

Reply:
Yeah, I think that would be a better approach